In [3]:
#EX1A

import numpy as np

def gauss_elimination(A, b):
    A = A.astype(float)
    b = b.astype(float)
    n = len(b)

   
    for i in range(n):
    
        max_row = np.argmax(abs(A[i:, i])) + i
        A[[i, max_row]] = A[[max_row, i]]
        b[[i, max_row]] = b[[max_row, i]]

        for j in range(i + 1, n):
            factor = A[j, i] / A[i, i]
            A[j, i:] -= factor * A[i, i:]
            b[j] -= factor * b[i]

    x = np.zeros(n)
    for i in range(n - 1, -1, -1):
        x[i] = (b[i] - np.dot(A[i, i+1:], x[i+1:])) / A[i, i]

    return x


In [4]:
#EX1B

A = np.array([
    [10, 7, 8, 7],
    [7, 5, 6, 5],
    [8, 6, 10, 9],
    [7, 5, 9, 10]
], dtype=float)

A_tilde = np.array([
    [10, 7, 8.1, 7.2],
    [7.08, 5.04, 6, 5],
    [8, 6, 9.98, 9],
    [6.99, 4.99, 9, 9.98]
], dtype=float)


b = np.array([32, 23, 33, 31], dtype=float)
b_tilde = np.array([32.1, 22.9, 33.1, 30.9], dtype=float)

# Lösungen
x = gauss_elimination(A.copy(), b.copy())
x_tilde = gauss_elimination(A_tilde.copy(), b_tilde.copy())

print("x =", x)
print("x_tilde =", x_tilde)

# Vergleich
print("Differenz ||x - x_tilde|| =", np.linalg.norm(x - x_tilde))


x = [1. 1. 1. 1.]
x_tilde = [  72.52731175 -117.79249826   31.59632648  -17.29942697]
Differenz ||x - x_tilde|| = 143.17408351098544


In [5]:
np.set_printoptions(precision=4, suppress=True)

def pretty(M, decimals=4):
    return np.round(M, decimals)


A_inv = np.linalg.inv(A)
A_tilde_inv = np.linalg.inv(A_tilde)

print("\nA^-1:\n", pretty(A_inv))
print("\nA_tilde^-1:\n", pretty(A_tilde_inv))


eig_A = np.linalg.eigvals(A)
eig_A_tilde = np.linalg.eigvals(A_tilde)

print("\nEigenwerte von A:", np.round(eig_A, 6))
print("Eigenwerte von A_tilde:", np.round(eig_A_tilde, 6))


def compute_norms(M):
    return tuple(
        round(float(np.linalg.norm(M, ord)), 4)
        for ord in [1, 2, np.inf]
    )

print("\nNormen von A (1,2,inf):", compute_norms(A))
print("Normen von A_tilde:", compute_norms(A_tilde))
print("Normen von A^-1:", compute_norms(A_inv))
print("Normen von A_tilde^-1:", compute_norms(A_tilde_inv))



A^-1:
 [[ 25. -41.  10.  -6.]
 [-41.  68. -17.  10.]
 [ 10. -17.   5.  -3.]
 [ -6.  10.  -3.   2.]]

A_tilde^-1:
 [[ 339.4225 -533.6026  134.2094  -98.5685]
 [-562.5467  885.1202 -223.0309  163.5284]
 [ 143.8088 -226.681    57.8798  -42.3784]
 [ -86.1457  135.5973  -34.6811   25.5905]]

Eigenwerte von A: [30.2887  3.8581  0.0101  0.8431]
Eigenwerte von A_tilde: [30.3783  3.7869  0.0008  0.834 ]

Normen von A (1,2,inf): (33.0, 30.2887, 33.0)
Normen von A_tilde: (33.08, 30.3789, 32.98)
Normen von A^-1: (136.0, 98.5217, 136.0)
Normen von A_tilde^-1: (1781.0011, 1307.2579, 1834.2262)


In [6]:
#EX2A:

import numpy as np

A = np.array([
    [10, 7, 8, 7],
    [7, 5, 6, 5],
    [8, 6, 10, 9],
    [7, 5, 9, 10]
], dtype=float)

b = np.array([32, 23, 33, 31], dtype=float)

A, b

(array([[10.,  7.,  8.,  7.],
        [ 7.,  5.,  6.,  5.],
        [ 8.,  6., 10.,  9.],
        [ 7.,  5.,  9., 10.]]),
 array([32., 23., 33., 31.]))

In [7]:
# Check symmetry
print("Symmetric:", np.allclose(A, A.T))

# Check positive definiteness via Cholesky
try:
    np.linalg.cholesky(A)
    print("Positive definite: True")
except np.linalg.LinAlgError:
    print("Positive definite: False")

Symmetric: True
Positive definite: True


In [8]:
#B:

def cholesky(A):
    n = A.shape[0]
    L = np.zeros_like(A)

    for i in range(n):
        for j in range(i + 1):
            s = 0
            for k in range(j):
                s += L[i][k] * L[j][k]

            if i == j:
                L[i][j] = np.sqrt(A[i][i] - s)
            else:
                L[i][j] = (A[i][j] - s) / L[j][j]

    return L

L = cholesky(A)
print("L =\n", L)

L =
 [[3.1623 0.     0.     0.    ]
 [2.2136 0.3162 0.     0.    ]
 [2.5298 1.2649 1.4142 0.    ]
 [2.2136 0.3162 2.1213 0.7071]]


In [9]:
#C:

y = np.linalg.solve(L, b)

x = np.linalg.solve(L.T, y)

print("Solution x =", x)

Solution x = [1. 1. 1. 1.]


In [10]:
# Check Ax ≈ b
print("Ax =", A @ x)
print("b  =", b)

Ax = [32. 23. 33. 31.]
b  = [32. 23. 33. 31.]


In [11]:
import time

#Ex 3

# Jacobi Methode A
def jacobi(A, b, tol=1e-4, max_iter=10000):
    n = len(b)
    x = np.zeros(n)
    x_new = np.zeros(n)
    
    start = time.time()
    
    for k in range(max_iter):
        for i in range(n):
            s = sum(A[i][j] * x[j] for j in range(n) if j != i)
            x_new[i] = (b[i] - s) / A[i][i]
        
        
        if np.linalg.norm(x_new - x, ord=np.inf) < tol:
            break
        
        x = x_new.copy()
    
    end = time.time()
    
    return x_new, k+1, end - start



x_jacobi, iter_j, time_j = jacobi(A, b)

print("Jacobi:")
print("Solution:", x_jacobi)
print("Iterations:", iter_j)
print("Time:", time_j)

C:\Users\10\AppData\Local\Temp\ipykernel_7036\4065327652.py:15: RuntimeWarning: overflow encountered in scalar add
  s = sum(A[i][j] * x[j] for j in range(n) if j != i)


Jacobi:
Solution: [-inf -inf -inf -inf]
Iterations: 10000
Time: 0.26509523391723633


In [12]:
# Gauss-Seidel Methode B
def gauss_seidel(A, b, tol=1e-4, max_iter=10000):
    n = len(b)
    x = np.zeros(n)
    
    start = time.time()
    
    for k in range(max_iter):
        x_old = x.copy()
        
        for i in range(n):
            s1 = sum(A[i][j] * x[j] for j in range(i))      # new values
            s2 = sum(A[i][j] * x_old[j] for j in range(i+1, n))  # old values
            x[i] = (b[i] - s1 - s2) / A[i][i]
        
        if np.linalg.norm(x - x_old, ord=np.inf) < tol:
            break
    
    end = time.time()
    
    return x, k+1, end - start

x_gs, iter_gs, time_gs = gauss_seidel(A, b)
print("Gauss-Seidel:")
print("Solution:", x_gs)
print("Iterations:", iter_gs)
print("Time:", time_gs)


Gauss-Seidel:
Solution: [1.0195 0.9678 1.008  0.9952]
Iterations: 1094
Time: 0.07318234443664551


Die Jacobi-Methode konvergiert für dieses System nicht, sondern divergiert, wie der Überlauf und die daraus resultierenden unendlichen Werte zeigen. Dies liegt daran, dass die Matrix nicht ausreichend diagonal dominant und schlecht konditioniert ist, worauf der in Aufgabe 1 gefundene sehr kleine Eigenwert hindeutet.

Im Gegensatz dazu konvergiert die Gauss-Seidel-Methode weiterhin zur korrekten Lösung und zeigt somit eine bessere Stabilität.

Die Cholesky-Zerlegung ist hier die effizienteste und zuverlässigste Methode, da sie die exakte Lösung ohne Iteration direkt berechnet und nicht von Konvergenzproblemen beeinträchtigt wird.

In [13]:
# Jacobi
start = time.time()
x_j, it_j, _ = jacobi(A, b)
t_j = time.time() - start

# Gauss-Seidel
start = time.time()
x_gs, it_gs, _ = gauss_seidel(A, b)
t_gs = time.time() - start

# Cholesky
start = time.time()
L = np.linalg.cholesky(A)
y = np.linalg.solve(L, b)
x_chol = np.linalg.solve(L.T, y)
t_chol = time.time() - start

print(it_j, t_j)
print(it_gs, t_gs)
print(t_chol)

C:\Users\10\AppData\Local\Temp\ipykernel_7036\4065327652.py:15: RuntimeWarning: overflow encountered in scalar add
  s = sum(A[i][j] * x[j] for j in range(n) if j != i)


10000 0.5066843032836914
1094 0.031664133071899414
0.0004394054412841797


Die Jacobi-Methode konvergiert für das gegebene System nicht. Sie erreicht die maximale Anzahl an Iterationen (10000) und führt zu einem Überlauf, was auf Divergenz hindeutet. Daher ist die Jacobi-Methode für dieses Problem nicht geeignet.

Die Gauss-Seidel-Methode konvergiert zur korrekten Lösung, erfordert jedoch eine relativ große Anzahl an Iterationen (1094), was im Vergleich zu direkten Methoden zu einer deutlich längeren Laufzeit führt.

Die Cholesky-Zerlegung ist deutlich effizienter. Sie berechnet die exakte Lösung in sehr kurzer Zeit (~0,00044 s) ohne Iteration.